sources/ghostdog/src/main.rs

```diff
@@ -125,6 +168,46 @@ fn parse_device_name(device_info: &[u8], path: String) -> Result<String> {
         .to_string())
 }
 
+/// Read a file into a SupportedDevicesConfiguration Enum
+fn read_supported_devices_file(path: PathBuf) -> Result<SupportedDevicesConfiguration> {
+    let mut supported_devices_file =
+        fs::File::open(&path).context(error::OpenFileSnafu { path: path.clone() })?;
+    let mut supported_devices_str = String::new();
+    supported_devices_file
+        .read_to_string(&mut supported_devices_str)
+        .context(error::ReadFileSnafu { path: path.clone() })?;
+    let device_configuration: SupportedDevicesConfiguration =
+        serde_json::from_str(supported_devices_str.as_str())
+            .context(error::ParseGpuDevicesFileSnafu {})?;
+    Ok(device_configuration)
+}
+
+/// Search the Open GPU Supported Devices File to determine if the Open GPU Driver should be used based upon PCI devices present
+fn find_preferred_driver() -> Result<String> {
+    let open_gpu_devices = read_supported_devices_file(OPEN_GPU_SUPPORTED_DEVICES_PATH.into())?;
+    let list_input = pciclient::ListDevicesParam::builder()
+        .vendor(NVIDIA_VENDOR_ID)
+        .build();
+    let present_devices =
+        pciclient::list_devices(list_input).context(error::ListPciDevicesSnafu {})?;
+
+    for input_device in present_devices.iter() {
+        match open_gpu_devices {
+            SupportedDevicesConfiguration::OpenGpu(ref device_list) => {
+                let formatted_device_id = format!("0x{}", input_device.device());
+                for supported_device in device_list.iter() {
+                    if supported_device.device_id == formatted_device_id {
+                        return Ok("open-gpu".to_string());
+                    }
+                }
+            }
+        }
+    }
```

Comment:
In a system with a large number of GPUs that weren't supported by the open driver, and a JSON file with a large number of different devices, we could end up doing a lot of extra work.

It would be more efficient to transform the open GPU device data into a hash set of device IDs. Then we would guarantee one pass through the list of supported devices, and one lookup per device.

Reply:
I tried putting these into `HashSet`s to ensure we didn't have negative performance even in cases where there are 16 devices and the runs all were taking between 5-6ms both as the Vec implementation as it stands vs the HashSet `contains()`. It stands to reason that HashSet's would be better so I can convert, but I don't think we gain a lot here with the ~700 devices listed in the current file, even in worst case.
```
    // Bogus PCI devices but "miss" to ensure there will be no match
    let present_devices: Vec<ListDevicesOutput> = vec![ListDevicesOutput {
                pci_slot: "00:00.0".to_string(),
                class: "0600".to_string(),
                vendor: "8086".to_string(),
                device: "1234".to_string(),
                program_interface: Some("00".to_string()),
                subsystem_vendor: Some("1d0f".to_string()),
                subsystem_device: Some("1237".to_string()),
                ..Default::default()
            };16];

    // reduce down to unique ids so we only iterate on uniqueness (down from 16 to 1 here)
    let unique_ids: HashSet<String> = present_devices.iter().map(|x| format!("0x{}", x.device()).clone()).collect();
    match open_gpu_devices {
            SupportedDevicesConfiguration::OpenGpu(ref device_list) => {
                let open_devices_list: HashSet<String> = device_list.iter().map(|x| x.device_id.clone()).collect();
                for input_device in unique_ids.iter() {
                    if open_devices_list.contains(input_device) {
                        return Ok("open-gpu".to_string());
                    }
            }
       }
    }
```
The compiler might be tricking me here and optimizing as well  (I did try creating the Vec of unique objects and still saw the same times for input devices) so I can test a bit more, but even so, 5ms runs for this when parsing the entire file is good considering the kernel module loads will take seconds. We might save more by saving the decision to a file to be checked before rerunning the deserialization and searching since this will be called several times by systemd.